src/model_state.jl

```diff
@@ -0,0 +1,150 @@
+"""
+    $TYPEDEF
+
+Represents the state of a "simulation" for a given `model`. `ModelState` consists of a
+`clock`, a `model`, and an initialized `StateVariables` data structure, as well as a `cache`
+for the timestepper and any relevant `inputs` provided by a corresponding `InputProvider`.
+The `ModelState` implements the `Oceananigans.AbstractModel` interface and can thus be
+treated as a "model" in `Oceananigans` `Simulation`s and output reading/writing utilities.
+"""
+struct ModelState{
+    NF,
+    Arch<:AbstractArchitecture,
+    Grid<:AbstractLandGrid{NF, Arch},
+    TimeStepper<:AbstractTimeStepper{NF},
+    Model<:AbstractModel{NF, Grid, TimeStepper},
+    StateVars<:AbstractStateVariables,
+    Inputs<:InputProvider
+} <: Oceananigans.AbstractModel{TimeStepper, Arch}
+    "The clock holding all information about the current timestep/iteration of a simulation"
+    clock::Clock
+
+    "Spatial discretization of the underlying `model`"
+    grid::Grid
+
+    "The type of model used for the simulation."
+    model::Model
+
+    "Input provider type"
+    inputs::Inputs
+
+    "Collection of all state variables defined on the simulation `model`."
+    state::StateVars
+end
+
+# Oceananigans.AbstractModel interface
+
+Base.time(state::ModelState) = state.clock.time
+
+Base.eltype(::ModelState{NF}) where {NF} = NF
+
+iteration(state::ModelState) = state.clock.iteration
+
+architecture(state::ModelState) = architecture(get_grid(state.model))
+
+timestepper(state::ModelState) = get_time_stepping(state.model)
+
+function update_state!(state::ModelState; compute_tendencies = true)
+    reset_tendencies!(state.state)
+    update_inputs!(state.inputs, state.clock)
+    compute_auxiliary!(state.state, state.model)
+    if compute_tendencies
+        compute_tendencies!(state.state, state.model)
+    end
+end
+
+# for now, just forward Oceananigans.time_step! to timestep!
+# consider renaming later...
+time_step!(state::ModelState, Δt; kwargs...) = timestep!(state, Δt)
+
+"""
+    $TYPEDEF
+
+Resets the simulation `clock` and calls `initialize!(state, model)` on the underlying model which
+should reset all state variables to their values as defiend by the model initializer.
+"""
+function initialize!(state::ModelState)
+    # TODO: reset other variables too?
```

Comment:
I am a bit confused, does the `initialize!`(::State, ::Model)` reset the `State` variables? I thought not. Because the doctoring is kind of implying that, but here you are wondering whether they should be reset separately.

Reply:
Am I misunderstanding this, or is this a rather big change, why would you compute the auxiliary variables before the timestep?